deep_qa/training/trainer.py

```diff
@@ -509,6 +515,95 @@ def _dataset_indexing_kwargs(self) -> Dict[str, Any]:
     # Protected methods - you CAN override these, if you want
     ###################
 
+    def _compile_parallel_model(self) -> DeepQaModel:
+        """
+        This method compiles a multi-gpu version of your model. This is done using data
+        parallelism, by making N copies of the model on the different GPU, all of which
+        share parameters. Gradients are updated synchronously, using the average gradient
+        from all of the outputs of the various models. This effectively allows you to scale
+        a model up to batch_sizes which cannot fit on a single GPU.
+
+        This method returns a "primary" copy of the model, which has had it's training
+        function which is run by Keras overridden to be a training function which trains
+        all of the towers of the model. The other towers never have their training functions
+        initialised or used and are completely hidden from the user. The returned model
+        can be serialised in the same way as any other model and has no dependency on
+        multiple gpus being available when it is loaded.
+
+        Returns
+        -------
+        The "primary" copy of the DeepQaModel, which holds the training function which
+        trains all of the copies of the model.
+        """
+        tower_models = []
+        tower_gradients = []
+        global_step = tensorflow.train.get_or_create_global_step()
+        train_loss = tensorflow.get_variable('train_loss', [],
+                                             initializer=tensorflow.constant_initializer(0.0),
+                                             trainable=False)
+
+        # Place a copy of the model on each GPU, each getting a slice of the batch.
+        for gpu_index in range(self.num_gpus):
+            with tensorflow.device(pin_variable_device_scope('/gpu:%d' % gpu_index)):
+                with tensorflow.name_scope('tower_%d' % gpu_index):
+                    # This is a new model object every time.
+                    model = self._build_model()
+                    # We are using the optimizer directly here, so we don't clutter
+                    # the graph creation by creating optimizers we don't use.
+                    compile_kwargs = self.__compile_kwargs()
+                    compile_kwargs['optimizer'] = None
+                    model.compile(self.__compile_kwargs())
+                    loss = model.total_loss
+                    tower_models.append(model)
+                    grads = self.optimizer.compute_gradients(loss)
+                    tower_gradients.append(grads)
+                    train_loss += loss
+
+        grads = average_gradients(tower_gradients)
+        train_operation = self.optimizer.apply_gradients(grads, global_step=global_step)
```

Comment:
Do you plan to support gradient clipping?  I've found it essential for training certain kinds of RNNs.  This would be the place to include it.